dot/parachain/prospective-parachains/active_leaves_update.go

```diff
@@ -0,0 +1,383 @@
+
+type BlockState interface {
+	GetHeader(hash common.Hash) (header *types.Header, err error)
+	GetRuntime(blockHash common.Hash) (instance runtime.Instance, err error)
+}
+
+// ProcessActiveLeavesUpdateSignal processes active leaves update signal
+func (pp *ProspectiveParachains) ProcessActiveLeavesUpdateSignal(
+	msg parachaintypes.ActiveLeavesUpdateSignal,
+) error {
+	// ensure there activated is not nil and it is not in the deactivated list (sanity check)
+	if msg.Activated != nil && !slices.Contains(msg.Deactivated, msg.Activated.Hash) {
+		tmpHeaderCache := make(map[common.Hash]*types.Header)
+
+		hash := msg.Activated.Hash
+
+		runtimeInstance, err := pp.blockState.GetRuntime(hash)
+		if err != nil {
+			return fmt.Errorf("getting runtime: %w", err)
+		}
+
+		claimQueue, err := runtimeInstance.ParachainHostClaimQueue()
+		if err != nil {
+			return fmt.Errorf("fetching claim queue: %w", err)
+		}
+
+		transposed := claimQueue.ToTransposed()
+
+		blockInfo, err := pp.fetchBlockInfo(tmpHeaderCache, hash)
+		if err != nil {
+			return fmt.Errorf("fetching block info: %w", err)
+		}
+
+		sessionIndex, err := runtimeInstance.ParachainHostSessionIndexForChild()
+		if err != nil {
+			return fmt.Errorf("requesting session index for child: %w", err)
+		}
+
+		ancestryLen, err := runtimeInstance.ParachainHostSchedulingLookAhead()
+		if err != nil {
+			logger.Warnf("fetching scheduling lookahead: %s", err.Error())
+			ancestryLen = DefaultSchedulingLookAhead
+		}
+
+		ancestryLen = primitives.SaturatingSub(ancestryLen, 1)
+
+		ancestry, err := pp.fetchAncestry(
+			hash,
+			ancestryLen,
+			sessionIndex,
+			tmpHeaderCache,
+		)
+		if err != nil {
+			return fmt.Errorf("fetching ancestry: %w", err)
+		}
+
+		ancestorsHashes := make([]string, len(ancestry))
+		ancestryBlockInfo := make([]relayChainBlockInfo, len(ancestry))
+		for idx, a := range ancestry {
+			ancestryBlockInfo[idx] = relayChainBlockInfo{
+				Hash:        a.Hash,
+				Number:      a.Number,
+				StorageRoot: a.StorageRoot,
+			}
+			ancestorsHashes[idx] = a.Hash.String()
+		}
+
+		var prevFragmentChains map[parachaintypes.ParaID]*fragmentChain
+		if len(ancestry) > 0 {
+			prevFragmentChains = pp.view.perRelayParent[ancestry[0].Hash].fragmentChains
+		}
+
+		fragmentChains := make(map[parachaintypes.ParaID]*fragmentChain)
+		for paraID, claimsByDepth := range transposed {
+			constraints, err := runtimeInstance.ParachainHostBackingConstraints(paraID)
+			if err != nil {
+				return fmt.Errorf("fetching backing constraints: %w", err)
+			}
+
+			if constraints == nil {
+				logger.Warnf("failed to get inclusion backing constraints")
+				continue
+			}
+
+			pendingAvailabilityCandidates, err := runtimeInstance.
+				ParachainHostCandidatesPendingAvailability(
+					paraID,
+				)
+			if err != nil {
+				return fmt.Errorf("fetching pending availability candidates: %w", err)
+			}
+
+			importablePending, err := pp.preprocessCandidatesPendingAvailability(
```

Comment:
what does it mean they are importable? Should this be something like pendingCandidates instead?

Reply:
not all pending candidates are importable, meaning they were pruned or already imported.